sdk/python/kfp/compiler/compiler_utils.py

```diff
@@ -427,14 +427,29 @@ def get_dependencies(
                 task2=task,
             )
 
-            # a task cannot depend on a task created in a for loop group since individual PipelineTask variables are reassigned after each loop iteration
+            # ParralelFor Check
+            for parent in task_name_to_parent_groups[upstream_task.name]:
+                parent = group_name_to_group.get(parent, None)
+                if isinstance(
+                        parent,
+                        tasks_group.ParallelFor) and task not in parent.tasks:
+                    raise RuntimeError(
+                        f'Task {task.name} cannot dependent on any task inside a ParralelFor'
+                    )
+
+            # Condition check
+            dependent_group = group_name_to_group.get(upstream_groups[0], None)
+            if isinstance(dependent_group, tasks_group.Condition):
+                raise RuntimeError(
+                    f'Task {task.name} cannot dependent on any task inside a Condition that is not a common ancestor of both tasks'
+                )
+
+            # ExitHandler check
             dependent_group = group_name_to_group.get(upstream_groups[0], None)
-            if isinstance(dependent_group,
-                          (tasks_group.ParallelFor, tasks_group.Condition,
-                           tasks_group.ExitHandler)):
+            if isinstance(dependent_group, tasks_group.ExitHandler):
                 raise RuntimeError(
-                    f'Task {task.name} cannot dependent on any task inside'
-                    f' the group: {upstream_groups[0]}.')
+                    f'Task {task.name} cannot dependent on any task inside a Exithandler that is not a common ancestor of both tasks'
```

Comment:
Apologies -- just remembered we discussed offline and this preserves existing behavior.

Reply:
Hey Connor, so would it be okay to mark this as resolved, since no action is required?

Reply:
Perhaps @chensun can offer a second opinion. @chensun, WDYT about `task` consuming from `inner_task` in the pipeline above?

Reply:
(Sorry for my slow response)

What if `inner_task` failed? Suppose the exit task should handle it gracefully and allow continuing execution of the remaining tasks, then `task` would reference a non-exists output, isn't that just like a conditional case?

We can translate the above example to Python code like:
```python
try:
    inner_task = identity(message=message)
    fail_op(message='Task failed.')
except:
    ...
finally:
    exit_task = identity(message='Exit handler has worked!')

task = identity(message=inner_task.output)
```
Would this makes sense in the Python programming context? Understand Python doesn't block such usage, but it feels to me the last line of `task = ...` could, and probably should be, moved into the `try` block instead. WDTY?

Reply:
Thanks for mapping this to normal Python -- that's very helpful.

I think the reason for not moving `task` into the `try` block is the same as for normal Python: it might catch exceptions unrelated to the one you're trying to capture. KFP's `ExitHandler` is analogous to a ["bare" `except`](https://www.flake8rules.com/rules/E722.html) in normal Python, so all the more reason to have the minimum required tasks in the `ExitHandler`.

My personal preference is to permit the above syntax. It valid from a control flow standpoint and I don't see much ambiguity for the author.

Reply:
Either way, I think we should leave the behavior as it currently stands for this PR and address later if we want to :)

Reply:
After some testing, I see that there's actually more to enabling this behavior than just unblocking it.

The reference to inner task in `identity(message=inner_task.output)` fails to compile, since we don't have logic to register the outputs from `inner_task` to the parent exit handler sub-DAG.

All the more reason to leave as is for this PR.